polo-bukkit/src/main/java/dev/dhdf/polo/bukkit/Main.java

```diff
@@ -94,6 +101,66 @@ public void broadcastMessage(String message) {
         this.getServer().broadcastMessage(message);
     }
 
+    @Override
+    public void kickPlayer(UUID uuid, String reason, String source) {
+        Player player = this.getServer().getPlayer(uuid);
+        if (player == null) {
+            Logger logger = getLogger();
+            logger.warning("No player to kick with UUID " + uuid.toString());
+            return;
+        }
+        getServer().getScheduler().runTask(this, new Runnable() {
+            public void run() {
+                player.kickPlayer(reason);
+                getServer().broadcast(source + " kicked " + player.getName() + " for " + reason, "matrix.kick.notify");
+            }
+        });
+    }
+
+    @Override
+    public void banPlayer(UUID uuid, String reason, String source) {
+        Logger logger = getLogger();
+        OfflinePlayer offlinePlayer = getServer().getOfflinePlayer(uuid);
+        if (offlinePlayer == null) {
+            logger.warning("No player to ban with UUID " + uuid.toString());
+            return;
+        }
+        BanList banlist = getServer().getBanList(BanList.Type.NAME);
+        if (banlist == null) {
+            logger.warning("No banlist found");
+            return;
+        }
+        banlist.addBan(offlinePlayer.getName(), reason, null, source);
```

Comment:
Should probably use UUID here
```suggestion
        banlist.addBan(offlinePlayer.getUniqueId().toString(), reason, null, source);
```

Reply:
Note I'm getting the banlist by name, since the docs appear to say that there isn't a UUID banlist. I'll have another play around

Reply:
oh, i see. passing UUIDs as strings works. I didn't think to try that. I'll clean these up to work primarily from the UUID.

Reply:
> oh, i see. passing UUIDs as strings works. I didn't think to try that. I'll clean these up to work primarily from the UUID.

actually (at least on papermc) it only seems to work because it gets converted to a name. Banning players who haven't been seen before in this way seems to be unstable (i've seen long delays and exceptions), so I think its simpler to just use the name from the offlinePlayer (with null check). It still seems to store the UUID in the banned-players.json file.

Reply:
You're correct about it not saying anything about UUIDs, but the CraftBukkit code as in use by Spigot [tries to parse the string as UUID](https://hub.spigotmc.org/stash/projects/SPIGOT/repos/craftbukkit/browse/src/main/java/org/bukkit/craftbukkit/CraftProfileBanList.java#98-109), and retrieves the `GameProfile` to ban using that UUID if it succeeds. Otherwise, it'll retrieve the `GameProfile` by name. The same holds for unbanning.

On the other hand, `addBan` may return `null` if no ban was added (e.g. if the player is not known to the server)

Reply:
thanks for the source link.

so it looks like my delays are coming from the appservice end when banning fake users that don't exist, since we ignore /user/xx requests from homeserver so it takes ages to pass the ban event on. with that fixed it behaves sensibly, so I'll ban and unban by UUID as you've suggested